nucypher/policy/orders.py

```diff
@@ -0,0 +1,282 @@
+
+                if len(ursulas_and_cfrags) >= self._threshold:
+                    self._complete_results[capsule] = ursulas_and_cfrags
+                else:
+                    for ursula in ursulas_and_cfrags:
+                        ursulas_to_contact.add(ursula)
+                    self._partial_results[capsule] = ursulas_and_cfrags
+
+            ursulas_pick_order = [ursula for ursula in self._ursulas_pick_order
+                                  if ursula not in ursulas_to_contact]
+            self._ursulas_pick_order = list(ursulas_to_contact) + ursulas_pick_order
+
+    def has_work_orders(self):
+        # Or raise an exception in `get_work_order()`?
+        return bool(self._ursulas_pick_order)
+
+    def get_work_order(self):
+        while self._ursulas_pick_order:
+            ursula_address = self._ursulas_pick_order.pop(0)
+            capsules = [capsule for capsule in self._capsules
+                        if capsule not in self._processed_capsules.get(ursula_address, set())]
+            if len(capsules) > 0:
+                return WorkOrder(ursula_address=ursula_address,
+                                 capsules=capsules)
+        raise RuntimeError("No Ursulas left")
+
+    def get_cfrags(self, capsule):
+        return list(self._complete_results[capsule].values())
+
+    def update(self, work_order, cfrags):
+
+        # In case several Ursulas gave us the same cfrag.
+        # TODO: Raise a better error here?
+        assert len(set(cfrags.values())) == len(cfrags)
+
+        for capsule, cfrag in cfrags.items():
+            self._partial_results[capsule][work_order.ursula_address] = cfrag
+            if len(self._partial_results[capsule]) >= self._threshold:
+                self._complete_results[capsule] = self._partial_results[capsule]
+                del self._partial_results[capsule]
+
+    def is_complete(self):
+        return len(self._complete_results) == len(self._capsules)
+
+    def fill_message_kits(self, message_kits):
+        results = self.get_results()
+        new_message_kits = []
+        for message_kit in message_kits:
+            if message_kit.capsule not in results:
+                new_message_kits.append(message_kit)
+            else:
+                cfrags = results[message_kit.capsule].values()
+                new_message_kit = message_kit.with_cfrags(cfrags)
+                new_message_kits.append(new_message_kit)
+        return new_message_kits
+
+    def get_results(self):
+        # keys in partial and complete results are guaranteed to be distinct
+        results = {capsule: dict(ursulas_and_cfrags) for capsule, ursulas_and_cfrags in self._partial_results.items()}
+        results.update({capsule: dict(ursulas_and_cfrags) for capsule, ursulas_and_cfrags in self._complete_results.items()})
+        return results
+
+
+class WorkOrder:
+
+    def __init__(self, ursula_address, capsules):
+        self.ursula_address = ursula_address
```

Comment:
Why do we only include `ursula_address` now instead of the entire Ursula?  This makes it much more possible to send a WorkOrder to a wrong Ursula, no?

Reply:
I don't see how, an address identifies an Ursula just as well as an Ursula object.

Reply:
The full `Ursula` has its `EncryptingPower` (making it possible only to encrypt for that Ursula) and ships with its server address.

Reply:
But we don't need that power in this context. Everything that had to be encrypted for the Ursula (kfrag) is already encrypted.